contracts/XINV.sol

```diff
@@ -0,0 +1,965 @@
+                uint96 dstRepNew = add96(dstRepOld, amount, "INV::_moveVotes: vote amount overflows");
+                _writeCheckpoint(dstRep, dstRepNum, dstRepOld, dstRepNew);
+            }
+        }
+    }
+
+    function _writeCheckpoint(address delegatee, uint32 nCheckpoints, uint96 oldVotes, uint96 newVotes) internal {
+      uint32 blockNumber = safe32(block.number, "INV::_writeCheckpoint: block number exceeds 32 bits");
+
+      if (nCheckpoints > 0 && checkpoints[delegatee][nCheckpoints - 1].fromBlock == blockNumber) {
+          checkpoints[delegatee][nCheckpoints - 1].votes = newVotes;
+      } else {
+          checkpoints[delegatee][nCheckpoints] = Checkpoint(blockNumber, newVotes);
+          numCheckpoints[delegatee] = nCheckpoints + 1;
+      }
+
+      emit DelegateVotesChanged(delegatee, oldVotes, newVotes);
+    }
+
+    function getChainId() internal pure returns (uint) {
+        uint256 chainId;
+        assembly { chainId := chainid() }
+        return chainId;
+    }
+
+    function safe32(uint n, string memory errorMessage) internal pure returns (uint32) {
+        require(n < 2**32, errorMessage);
+        return uint32(n);
+    }
+
+    function add96(uint96 a, uint96 b, string memory errorMessage) internal pure returns (uint96) {
+        uint96 c = a + b;
+        require(c >= a, errorMessage);
+        return c;
+    }
+
+    function sub96(uint96 a, uint96 b, string memory errorMessage) internal pure returns (uint96) {
+        require(b <= a, errorMessage);
+        return a - b;
+    }
+}
+
+contract TimelockEscrow {
+    using SafeMath for uint;
+
+    address public underlying;
+    address public governance;
+    address public market;
+    uint public duration = 14 days;
+    mapping (address => EscrowData) public pendingWithdrawals;
+
+    struct EscrowData {
+        uint timestamp;
+        uint amount;
+    }
+
+    constructor(address underlying_, address governance_) public {
+        underlying = underlying_;
+        governance = governance_;
+        market = msg.sender;
+    }
+
+    // set to 0 to send funds directly to users
+    function _setEscrowDuration(uint duration_) public {
+        require(msg.sender == governance, "only governance can set escrow duration");
+        duration = duration_;
+    }
+
+    function _setGov(address governance_) public {
+        require(msg.sender == governance, "only governance can set its new address");
+        governance = governance_;
+    }
+
+    /**
+     * @notice assumes funds were already sent to this contract by the market. Resets escrow timelock on each withdrawal
+     */
+    function escrow(address user, uint amount) public {
+        require(msg.sender == market, "only market can escrow");
+        if(duration > 0) {
+            EscrowData memory withdrawal = pendingWithdrawals[user];
+            pendingWithdrawals[user] = EscrowData({
+                timestamp: block.timestamp,
```

Comment:
So if a user deposits and redeems every 13 days, they could never withdraw it from the time lock. It'd be their own fault, but still it seems a bit strange.

Why the time reset, and not a timestamp per received escrow?

Reply:
- Easier to map each user address to a withdrawal
- Penalty for frequent withdrawals in exchange for high rewards